Add option to exit when stdin ends

michaeljones opened the pull request:
This is a first draft/approach to the `--stdin-quit` functionality raised in #440. I am not experienced with the code base and I'm not very experienced with Rust. I'm happy to change any and all of this. I just hope it isn't so wrong as to be completely useless.

Important to note that this first draft always quits on stdin ending regardless of the flag usage. I couldn't see how the runtime config is used to influence the behaviour as it is not clearly used in either the event generation code that I inspected or the `on_action` code but I assume I'm missing something. It would be reasonably easy to pass to either to ensure either the `Keyboard::Eof` tagged event isn't generated without the flag or that the `Stop/Exit` outcome isn't generated for the event without the flag.

### Description

The watchexec cli gains a '--stdin-quit' flag which will cause the watchexec to end & exit when stdin for watchexec closes. This is to make it easier to use watchexec as a subprocess and make sure that it ends with the parent process ends, provided the parent process controls the stdin to the watchexec process.

We model the end of stdin as a keyboard event as it is possible to send 'eof' from a terminal using 'ctrl-d'. We keep reading lines from stdin and if we get back 0 bytes we assume that means that the stream has closed. This is a technique I have used elsewhere but I'm not that experienced so there might be a much better way to do it.

### Manual testing

A simple cargo run with ctrl-d to check that stdin closing ends the watching. I'm not sure if there are better ways.

```
$ cargo run -q -- "echo hello"
hello
hello
$ # Process has ended due to ctrl-d input
```

### Checklist

- [x] Ignore Eof event when `--stdin-quit` is not set
- [ ] Add unit tests as appropriate
- [x] Improve naming of field and functions
- [x] Add documentation to fix clippy warnings
- [x] Add man page entry

passcod commented on crates/lib/src/watchexec.rs:
```diff
@@ -105,11 +105,12 @@ impl Watchexec {
 			let fs = SubTask::spawn("fs", fs::worker(fs_r, er_s.clone(), ev_s.clone()));
 			let signal =
 				SubTask::spawn("signal", signal::source::worker(er_s.clone(), ev_s.clone()));
+			let keyboard = SubTask::spawn("keyboard", keyboard::worker(er_s.clone(), ev_s.clone()));
```
see how the fs worker has three arguments and the first is a watch channel of the fs working data? that's how you get the config in there

passcod commented on crates/lib/src/keyboard.rs:
```diff
@@ -0,0 +1,78 @@
+	errors: mpsc::Sender<RuntimeError>,
+	events: priority::Sender<Event, Priority>,
+) -> Result<(), CriticalError> {
+	imp_worker(errors, events).await
```
config handling: this function essentially should do a loop that watches the working data. if the bool changes to true, start up an `imp_worker` task and keep a handle, if it changes to false, cancel the task (or send a cancel, we want to drop the stdin held within the task and not leave it dangling), otherwise do nothing and wait more.

michaeljones replied:
I haven't tackled this yet. The code now included something that watches for reconfigurations but it won't interrupt the stdin monitoring to do it. I'll try to look at this soon.

michaeljones replied:
I've made an attempt at this. I think it is reasonable but wouldn't be surprised if there was a nicer way to achieve it.

passcod commented on crates/cli/src/config/runtime.rs:
```diff
@@ -120,6 +125,13 @@ pub fn runtime(args: &ArgMatches) -> Result<RuntimeConfig> {
 			return fut;
 		}
 
+		let tags: Vec<_> = action.events.iter().flat_map(|e| e.tags.clone()).collect();
```
simpler to do a `action.events.find(|e| e.tags.contains...)` or something like it

michaeljones replied:
Good point. I've updated that.

passcod reviewed (changes requested):
that's a great start! I've added a bunch of notes

michaeljones commented on crates/lib/src/config.rs:
```diff
@@ -62,6 +65,11 @@ impl RuntimeConfig {
 		self
 	}
 
+	pub fn keyboard(&mut self) -> &mut Self {
```
Changed.

michaeljones commented on crates/lib/src/keyboard.rs:
```diff
@@ -0,0 +1,78 @@
+	pub eof: bool,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Done.

michaeljones commented:
Thanks for the review. I can see to further update the code so the the config change will interrupt the stdin wait loop. I'll give it a go soon.

michaeljones commented:
Ready for re-review. Still needs documentation and tests if we want them. And any changes you feel are necessary.

passcod commented:
That looks great! With documentation, this is mergeable

michaeljones commented:
Added doc comments where needed & a man page entry. Also added some error handling as we were getting build warnings about not handling the potential error from the oneshot channel `.send`. Not sure if it looks reasonable it should be made more specific.

passcod commented:
A couple of clippy failures. Also that section could be a match instead, I think.

michaeljones commented on crates/lib/src/error/runtime.rs:
```diff
@@ -50,6 +50,11 @@ pub enum RuntimeError {
+	/// Events from the keyboard event source
+	#[error("keyboard watcher error")]
+	#[diagnostic(code(watchexec::runtime::keyboard_watcher))]
+	KeyboardWatcher,
```
Would it be sensible to have some kind of `kind` field on this so that we can disambiguate any future keyboard errors that might be identified without breaking builds?

passcod replied:
Sorry for the delay! Yeah what would be good is to do like the FsWatcher error, minus the `kind` field which isn't needed.

michaeljones commented:
> A couple of clippy failures. Also that section could be a match instead, I think.

Fix the clippy issues and refactored the multi-if into a match (if that is the one you were talking about.) 

I've also raised a question about the keyboard error and whether it should be more extensible?

michaeljones commented:
Also if there are any changes you'd like and you're happy to do yourself then, obviously, place change as you want. Sometimes it is faster to just do that.

michaeljones commented on crates/lib/src/keyboard.rs:
```diff
@@ -0,0 +1,125 @@
+	let mut send_close = None;
+	while working.changed().await.is_ok() {
+		let watch_for_eof = { working.borrow().eof };
+		match (watch_for_eof, send_close.take()) {
```
Hmm, maybe the refactor to a match makes this 'take' problematic? Might need to go back.

michaeljones replied:
I've refactored to do the 'take' in an inner if-let. Not sure if that is necessary but I think it is.

Please make any changes that you feel are necessary and merge if you get to a happy state. Though the question about the Keyboard error is still outstanding for the moment.

passcod commented on crates/lib/src/error/runtime.rs:
```diff
@@ -50,6 +50,11 @@ pub enum RuntimeError {
+	/// Events from the keyboard event source
+	#[error("keyboard watcher error")]
+	#[diagnostic(code(watchexec::runtime::keyboard_watcher))]
+	KeyboardWatcher,
```
```suggestion
	KeyboardWatcher {
		/// The underlying error.
		#[source]
		err: super::KeyboardWatcherError,
	},
```

michaeljones commented:
@passcod  - thanks for the review and the suggestion. I've incorporated your change and updated the rest of the code to use it. 

I'm not aware of any other changes that need to happen but I'm happy to change anything you deem necessary.

michaeljones commented:
Ace! Happy to see it merged. I hope it proves useful for people.
